<!DOCTYPE html>
<html>
<head>
    <meta http-eqiv='content-type' content='text/html;charset=utf-8'>
    <title>bg7 script</title>
    <link rel=stylesheet href="http://resources.ohnosequences.com/css/shocco.css">
    <link rel=stylesheet href="http://resources.ohnosequences.com/css/solarized.css">
</head>
<body>
<div id=container>
    <div id=background></div>
    <table cellspacing=0 cellpadding=0>
    <thead>
      <tr>
        <th class=docs><h1>bg7 script</h1></th>
        <th class=code></th>
      </tr>
    </thead>
    <tbody>
        <tr><td class='docs'><h1>bg7</h1>
<h3>requirements</h3>
<p>this script requires a standard bg7 install, with $BG7_HOME set.
Inside there, you should have</p>
</td><td class=code><div class=highlight><pre>
<span class="c">#!/bin/sh</span>

</pre></div></td></tr><tr><td class=docs>
<p>standard stuff. We should also be using <code>set -o nounset</code>. This prevents you from using unset variables.</p>
<p>exit at the point something's not working</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">set</span> -e

</pre></div></td></tr><tr><td class=docs>
<p>usage function. This returns the usage message below. Note that it is okay to echo multiline strings.</p>
</td><td class=code><div class=highlight><pre>
usage <span class="o">()</span>
<span class="o">{</span>
<span class="nb">echo</span> <span class="s2">&quot;</span>
<span class="s2">runs bg7 version 0.9. Params are:</span>

<span class="s2">  -n | --project_name: the name for your project. Defaults to ${default_name}</span>
<span class="s2">  -g | --genome: path to FASTA file with the sequences you want to annotate. It defaults to </span>
<span class="s2">                 &lt;name&gt;-genome.fasta</span>
<span class="s2">  -p | --proteins: FASTA file with ref proteins; defaults to &lt;name&gt;-proteins.fasta</span>
<span class="s2">  -r | --rnas: FASTA file with ref RNAs; defaults to &lt;name&gt;-rnas.fasta</span>
<span class="s2">  -c | --genetic_code: the genetic code you want to use; defaults to &lt;name&gt;-genetic_code</span>
<span class="s2">  -o | --out: the folder you want bg7 output to go; defaults to where you&#39;re running the </span>
<span class="s2">              program (this would be ${PWD} now)</span>
<span class="s2">  -G | --genbank_data: xml with extra data for your project needed for .gbk output (things like </span>
<span class="s2">                       locus name, Gen bank division tag, etc). There&#39;s a template for this at</span>
<span class="s2">                       ${BG7_HOME}/resources/</span>
<span class="s2">                        </span>
<span class="s2">  -h | --help: this message</span>

<span class="s2">&quot;</span> &gt;&amp;2
<span class="o">}</span>

</pre></div></td></tr><tr><td class=docs>
<p>default project name.</p>
</td><td class=code><div class=highlight><pre>
<span class="nv">default_name</span><span class="o">=</span>bg7project

</pre></div></td></tr><tr><td class=docs>
<p>params vars. This will match the params that will be used for execution, upon input param parsing etc.</p>
</td><td class=code><div class=highlight><pre>
<span class="nv">name</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
<span class="nv">proteins</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
<span class="nv">genome</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
<span class="nv">rnas</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
<span class="nv">genetic_code</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
<span class="nv">output_folder</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
<span class="nv">genbank_data</span><span class="o">=</span><span class="s2">&quot;&quot;</span>

</pre></div></td></tr><tr><td class=docs>
<h2>parse input</h2>
<p>Now, we will parse params, override corresponding param var after test, and make some fairly basic sanity checks.</p>
</td><td class=code><div class=highlight><pre>

</pre></div></td></tr><tr><td class=docs>
<p>no param then usage</p>
</td><td class=code><div class=highlight><pre>
<span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span>usage
  <span class="nb">exit </span>1
<span class="k">fi</span>
</pre></div></td></tr><tr><td class=docs>
<p>iterate through input params, using shift. Fragile, but it works!</p>
</td><td class=code><div class=highlight><pre>
<span class="k">while</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> !<span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">do</span>
<span class="k">  case</span> <span class="nv">$1</span> in
    -n | --project_name <span class="o">)</span>  <span class="nb">shift</span>
<span class="nb">      </span><span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>
        <span class="k">then</span>
<span class="k">          </span><span class="nb">echo</span> <span class="s2">&quot;empty project name parameter!&quot;</span>
          <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      </span><span class="nv">name</span><span class="o">=</span><span class="s2">&quot;$1&quot;</span>
    ;;
    -g | --genome <span class="o">)</span> <span class="nb">shift</span>
<span class="nb">      </span><span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;empty genome file parameter!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      if</span> <span class="o">[</span> ! -f <span class="nv">$1</span> <span class="o">]</span>; <span class="k">then </span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;genome file provided does not exist!&quot;</span>
        <span class="nb">exit </span>1  
      <span class="k">fi</span>
<span class="k">      </span><span class="nv">genome</span><span class="o">=</span><span class="s2">&quot;$1&quot;</span>     													
    ;;
    -G | --genbank_data <span class="o">)</span> <span class="nb">shift</span>
<span class="nb">      </span><span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;empty genbank data file parameter!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      if</span> <span class="o">[</span> ! -f <span class="nv">$1</span> <span class="o">]</span>; <span class="k">then </span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;genbank data file provided does not exist!&quot;</span>
        <span class="nb">exit </span>1  
      <span class="k">fi</span>
<span class="k">      </span><span class="nv">genbank_data</span><span class="o">=</span><span class="s2">&quot;$1&quot;</span>                               
    ;;
    -p | --proteins<span class="o">)</span>  <span class="nb">shift</span>
<span class="nb">      </span><span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;empty proteins file parameter!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      if</span> <span class="o">[</span> ! -f <span class="nv">$1</span> <span class="o">]</span>; <span class="k">then </span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;proteins file provided does not exist!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      </span><span class="nv">proteins</span><span class="o">=</span><span class="nv">$1</span>
    ;;
    -r | --rnas<span class="o">)</span>  <span class="nb">shift</span>
<span class="nb">      </span><span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;empty RNAs file parameter!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      if</span> <span class="o">[</span> ! -f <span class="nv">$1</span> <span class="o">]</span>; <span class="k">then </span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;RNAs file provided does not exist!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      </span><span class="nv">rnas</span><span class="o">=</span><span class="nv">$1</span>
    ;;
    -c | --genetic_code<span class="o">)</span>  <span class="nb">shift</span>
<span class="nb">      </span><span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;empty genetic code file parameter!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      if</span> <span class="o">[</span> ! -f <span class="nv">$1</span> <span class="o">]</span>; <span class="k">then </span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;genetic code file provided does not exist!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      </span><span class="nv">genetic_code</span><span class="o">=</span><span class="nv">$1</span>
    ;;
    -h | --help <span class="o">)</span>           usage
      <span class="nb">exit </span>1
    ;;
    -o | --out<span class="o">)</span>  <span class="nb">shift</span>
<span class="nb">      </span><span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$1&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;empty output folder!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      if</span> <span class="o">[</span> ! -d <span class="nv">$1</span> <span class="o">]</span>; <span class="k">then </span>
<span class="k">        </span><span class="nb">echo</span> <span class="s2">&quot;output folder provided does not exist!&quot;</span>
        <span class="nb">exit </span>1
      <span class="k">fi</span>
<span class="k">      </span><span class="nv">output_folder</span><span class="o">=</span><span class="sb">`</span>dirname <span class="nv">$1</span><span class="sb">`</span>/<span class="sb">`</span>basename <span class="nv">$1</span><span class="sb">`</span>
    ;;
    -h | --help <span class="o">)</span>           usage
      <span class="nb">exit </span>1
    ;;
    * <span class="o">)</span>                     usage
    <span class="nb">exit </span>1
    <span class="k">esac</span>
<span class="k">    </span><span class="nb">shift</span>
<span class="k">done</span>
</pre></div></td></tr><tr><td class=docs>
<h4>check and set defaults</h4>
<p>now, we will build defaults, if needed. First thing to check is if <code>$name</code> has been set; if not, just use the default.</p>
</td><td class=code><div class=highlight><pre>
<span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$name&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span><span class="nv">name</span><span class="o">=</span><span class="nv">$default_name</span>
  <span class="nb">echo</span> <span class="s2">&quot;you didn&#39;t provide a name for your project, will use default: ${name}&quot;</span>
<span class="k">fi</span>
</pre></div></td></tr><tr><td class=docs>
<p>Once we have <code>$name</code> solved, we can continue checking the rest of params. As some of them are relative to <code>$name</code>, we need to do this once <code>$name</code> is known.</p>
<p>check proteins file. Note that we need to check <em>again</em> for file existence.</p>
</td><td class=code><div class=highlight><pre>
<span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$proteins&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span><span class="nv">proteins</span><span class="o">=</span><span class="s2">&quot;${name}-proteins.fasta&quot;</span>
  <span class="k">if</span> <span class="o">[</span> ! -f <span class="s2">&quot;$proteins&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">    </span><span class="nb">echo</span> <span class="s2">&quot;no proteins file at: ${proteins}&quot;</span>
    <span class="nb">echo</span> <span class="s2">&quot;will have to leave now&quot;</span>
    <span class="nb">exit </span>1
  <span class="k">fi</span>
<span class="k">fi</span>
</pre></div></td></tr><tr><td class=docs>
<p>check RNAs file, same as for proteins</p>
</td><td class=code><div class=highlight><pre>
<span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$rnas&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span><span class="nv">rnas</span><span class="o">=</span><span class="s2">&quot;${name}-rnas.fasta&quot;</span>
  <span class="k">if</span> <span class="o">[</span> ! -f <span class="s2">&quot;$rnas&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">    </span><span class="nb">echo</span> <span class="s2">&quot;no RNAs file at: ${rnas}&quot;</span>
    <span class="nb">echo</span> <span class="s2">&quot;will have to leave now&quot;</span>
    <span class="nb">exit </span>1
  <span class="k">fi</span>
<span class="k">fi</span>
</pre></div></td></tr><tr><td class=docs>
<p>check genome seqs file...</p>
</td><td class=code><div class=highlight><pre>
<span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$genome&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span><span class="nv">genome</span><span class="o">=</span><span class="s2">&quot;${name}-genome.fasta&quot;</span>
  <span class="k">if</span> <span class="o">[</span> ! -f <span class="s2">&quot;$genome&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">    </span><span class="nb">echo</span> <span class="s2">&quot;no genome sequences file: ${genome}&quot;</span>
    <span class="nb">echo</span> <span class="s2">&quot;will have to leave now&quot;</span>
    <span class="nb">exit </span>1
  <span class="k">fi</span>
<span class="k">fi</span>
</pre></div></td></tr><tr><td class=docs>
<p>here we use the default bacterial genetic code. It'd be nice to add a link to <em>what</em> genetic code is this one.</p>
</td><td class=code><div class=highlight><pre>
<span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$genetic_code&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span><span class="nv">genetic_code</span><span class="o">=</span><span class="s2">&quot;${name}-genetic_code&quot;</span>
  <span class="k">if</span> <span class="o">[</span> ! -f <span class="s2">&quot;$genetic_code&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">    </span><span class="nb">echo</span> <span class="s2">&quot;no genetic code file: ${genetic_code}&quot;</span>
    <span class="nb">echo</span> <span class="s2">&quot;using default from: ${BG7_HOME}/resources/genetic_code&quot;</span>
    <span class="nv">genetic_code</span><span class="o">=</span><span class="s2">&quot;$BG7_HOME/resources/genetic_code&quot;</span>
  <span class="k">fi</span>
<span class="k">fi</span>
</pre></div></td></tr><tr><td class=docs>
<p>check output folder; if there's none, we'll try to use <code>$PWD</code>.</p>
</td><td class=code><div class=highlight><pre>
<span class="k">if</span> <span class="o">[</span> <span class="s2">&quot;$output_folder&quot;</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span><span class="nb">echo</span> <span class="s2">&quot;you didn&#39;t provide an output folder, will try to use ${PWD}&quot;</span>
  <span class="nv">output_folder</span><span class="o">=</span><span class="s2">&quot;${PWD}&quot;</span>
<span class="k">fi</span>

</pre></div></td></tr><tr><td class=docs>
<p>everything looks ok. let's create output folder</p>
</td><td class=code><div class=highlight><pre>
mkdir -p <span class="nv">$output_folder</span>

</pre></div></td></tr><tr><td class=docs>
<p>log params</p>
</td><td class=code><div class=highlight><pre>
<span class="nv">param_log</span><span class="o">=</span><span class="s2">&quot;params.log&quot;</span>
<span class="nb">echo</span> <span class="s2">&quot;logging your params to ${output_folder}/${param_log}&quot;</span>
touch <span class="nv">$output_folder</span>/params.log
<span class="nb">echo</span> <span class="s2">&quot;--project_name ${name}</span>
<span class="s2">--genome ${genome}</span>
<span class="s2">--proteins ${proteins}</span>
<span class="s2">--rnas ${rnas}</span>
<span class="s2">--genetic_code ${genetic_code}</span>
<span class="s2">--genbank_data ${genbank_data}</span>
<span class="s2">--output_folder ${output_folder}</span>
<span class="s2">&quot;</span> &gt; <span class="k">${</span><span class="nv">output_folder</span><span class="k">}</span>/<span class="k">${</span><span class="nv">param_log</span><span class="k">}</span>

</pre></div></td></tr><tr><td class=docs>
<h2>BLAST runs</h2>
<p>here we run blast for RNAs and proteins.
get number of processors; you need to set this manually for BLAST</p>
</td><td class=code><div class=highlight><pre>
<span class="nv">cores</span><span class="o">=</span><span class="k">$(</span>grep -c processor /proc/cpuinfo<span class="k">)</span>
<span class="nb">echo</span> <span class="s2">&quot;it looks like this machine has ${cores} cores, will use that for blast settings&quot;</span>
<span class="nb">echo</span> <span class="s2">&quot;cores used for BLAST: ${cores}</span>
<span class="s2">&quot;</span> &gt; <span class="k">${</span><span class="nv">output_folder</span><span class="k">}</span>/<span class="k">${</span><span class="nv">param_log</span><span class="k">}</span>
</pre></div></td></tr><tr><td class=docs>
<h3>RNAs vs contigs run</h3>
<p>first, we run RNAs vs contigs</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">cd</span> <span class="nv">$output_folder</span>

</pre></div></td></tr><tr><td class=docs>
<p>output vars</p>
</td><td class=code><div class=highlight><pre>
<span class="nv">rnasVsContigsSuffix</span><span class="o">=</span><span class="s2">&quot;_RNA_blastn.xml&quot;</span>
<span class="nv">rnasVsContigsOutputName</span><span class="o">=</span><span class="s2">&quot;${name}${rnasVsContigsSuffix}&quot;</span>
<span class="nv">rnasVsContigsOutputPath</span><span class="o">=</span><span class="s2">&quot;${output_folder}/${rnasVsContigsOutputName}&quot;</span>

</pre></div></td></tr><tr><td class=docs>
<p>creating blast db for RNAs</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">echo</span> <span class="s2">&quot;creating RNAs blast db&quot;</span>
makeblastdb -in <span class="s2">&quot;${rnas}&quot;</span> -dbtype <span class="s2">&quot;nucl&quot;</span> -out <span class="s2">&quot;${name}.rnas.db&quot;</span> -logfile <span class="s2">&quot;${name}.rnas.db.log&quot;</span>
</pre></div></td></tr><tr><td class=docs>
<p>run blastn</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">echo</span> <span class="s2">&quot;running blastn: RNAs vs genome sequence&quot;</span>
blastn -query <span class="s2">&quot;${genome}&quot;</span> -db <span class="s2">&quot;${name}.rnas.db&quot;</span> -outfmt <span class="s2">&quot;5&quot;</span> -evalue <span class="s2">&quot;1e-20&quot;</span> -max_target_seqs <span class="s2">&quot;1&quot;</span> -out <span class="s2">&quot;${rnasVsContigsOutputName}&quot;</span> -num_threads <span class="s2">&quot;${cores}&quot;</span>

</pre></div></td></tr><tr><td class=docs>
<p>check end element etc</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">echo</span> <span class="s2">&quot;basic blast results check&quot;</span>
<span class="nv">rnaBlastOk</span><span class="o">=</span><span class="k">$(</span>tail -n 1 <span class="k">${</span><span class="nv">rnasVsContigsOutputPath</span><span class="k">}</span> | grep -c <span class="s1">&#39;&lt;/BlastOutput&gt;&#39;</span><span class="k">)</span>
<span class="k">if</span> <span class="o">[</span> ! <span class="s2">&quot;${rnaBlastOk}&quot;</span> <span class="o">=</span> 1 <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span><span class="nb">echo</span> <span class="s2">&quot;looks like the RNA vs contigs blast didn&#39;t work&quot;</span>
  <span class="nb">exit </span>1
<span class="k">fi</span>
</pre></div></td></tr><tr><td class=docs>
<p>remove first two lines, I don't know why</p>
</td><td class=code><div class=highlight><pre>
sed <span class="s1">&#39;2d&#39;</span> <span class="nv">$rnasVsContigsOutputPath</span> &gt; <span class="nv">$rnasVsContigsOutputPath</span>.tmp
mv <span class="nv">$rnasVsContigsOutputPath</span>.tmp <span class="nv">$rnasVsContigsOutputPath</span>


</pre></div></td></tr><tr><td class=docs>
<h3>proteins vs contigs blast</h3>
<p>here we run tblastn against a blast db consisting in the ref proteins. Due to the way <code>makeblastdb</code> works, we need to run everything from the output folder.</p>
</td><td class=code><div class=highlight><pre>

</pre></div></td></tr><tr><td class=docs>
<h4>blast db creation</h4>
<p>go into output folder</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">cd</span> <span class="nv">$output_folder</span>
</pre></div></td></tr><tr><td class=docs>
<p>creating blast db for proteins</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">echo</span> <span class="s2">&quot;creating blast db with genome sequences&quot;</span>
makeblastdb -in <span class="s2">&quot;${genome}&quot;</span> -dbtype nucl -out <span class="s2">&quot;${name}.genome.db&quot;</span> -logfile <span class="s2">&quot;${name}.genome.db.log&quot;</span>

</pre></div></td></tr><tr><td class=docs>
<p>output file vars, needed to adhere to file naming conventions</p>
</td><td class=code><div class=highlight><pre>
<span class="nv">proteinsVsContigsSuffix</span><span class="o">=</span><span class="s2">&quot;_proteins_tBLASTn.xml&quot;</span>
<span class="nv">proteinsVsContigsOutName</span><span class="o">=</span><span class="nv">$name$proteinsVsContigsSuffix</span>
<span class="nv">proteinsVsContigsOutPath</span><span class="o">=</span><span class="nv">$output_folder</span>/<span class="nv">$proteinsVsContigsOutName</span>

</pre></div></td></tr><tr><td class=docs>
<h4>tblastn run</h4>
<p>run tblastn; note that we enclose everything in double-quotes.</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">echo</span> <span class="s2">&quot;running tblastn: proteins vs genome sequence&quot;</span>
tblastn -query <span class="s2">&quot;${proteins}&quot;</span> -db <span class="s2">&quot;${name}.genome.db&quot;</span> -outfmt <span class="s2">&quot;5&quot;</span> -evalue <span class="s2">&quot;1e-30&quot;</span> -max_target_seqs <span class="s2">&quot;1&quot;</span> -out <span class="s2">&quot;${proteinsVsContigsOutName}&quot;</span> -num_threads <span class="s2">&quot;${cores}&quot;</span>

</pre></div></td></tr><tr><td class=docs>
<h4>check tblastn results</h4>
<p>here we do a basic check, to see if the tblastn results look good. What we do is checking that the end line has the corresponding closing tag for BLAST xml output. If not, inform and leave.</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">echo</span> <span class="s2">&quot;basic blast results check&quot;</span>
<span class="nv">proteinBlastOk</span><span class="o">=</span><span class="sb">`</span>tail -n 1 <span class="k">${</span><span class="nv">proteinsVsContigsOutPath</span><span class="k">}</span> | grep -c <span class="s1">&#39;&lt;/BlastOutput&gt;&#39;</span><span class="sb">`</span>
<span class="k">if</span> <span class="o">[</span> ! <span class="s2">&quot;${proteinBlastOk}&quot;</span> <span class="o">=</span> 1 <span class="o">]</span>; <span class="k">then</span>
<span class="k">  </span><span class="nb">echo</span> <span class="s2">&quot;looks like tblastn proteins vs contigs didn&#39;t work&quot;</span>
  <span class="nb">exit </span>1
<span class="k">fi</span>
</pre></div></td></tr><tr><td class=docs>
<p>remove first line, I don't know why. It looks like jdom has some ugly bug when parsing an xml file with a link to a nonexistant DTD. In this case, removing the first 2 lines from the xml will make things work (no DTD declaration).</p>
</td><td class=code><div class=highlight><pre>
sed <span class="s1">&#39;2d&#39;</span> <span class="nv">$proteinsVsContigsOutPath</span> &gt; <span class="nv">$proteinsVsContigsOutPath</span>.tmp
mv <span class="nv">$proteinsVsContigsOutPath</span>.tmp <span class="nv">$proteinsVsContigsOutPath</span>

</pre></div></td></tr><tr><td class=docs>
<p>copy files to conform to strange requirements</p>
</td><td class=code><div class=highlight><pre>
cp <span class="nv">$genome</span> <span class="nv">$output_folder</span>/<span class="k">${</span><span class="nv">name</span><span class="k">}</span>_sequences.fna
cp <span class="nv">$proteins</span> <span class="nv">$output_folder</span>/<span class="k">${</span><span class="nv">name</span><span class="k">}</span>_ReferenceProteins.fasta
cp <span class="nv">$genetic_code</span> <span class="nv">$output_folder</span>/genetic_code.txt
cp <span class="nv">$genbank_data</span> <span class="nv">$output_folder</span>/<span class="k">${</span><span class="nv">name</span><span class="k">}</span>_GenBankExternalData.xml

</pre></div></td></tr><tr><td class=docs>
<h2>create executions.xml</h2>
<p>we have everything set, now we need to create the xml file which defines the execution, named (somewhat unexpectedly) <code>executions.xml</code>.
We'll do that through a HERE document; not the most orthodox way to create an xml file, but again, <em>it works</em>.</p>
</td><td class=code><div class=highlight><pre>
cat &gt; <span class="nv">$output_folder</span>/executions.xml <span class="s">&lt;&lt; EOF</span>
<span class="s">&lt;scheduled_executions&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.PredictGenes&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_proteins_tBLASTn.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_sequences.fna&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_PredictedGenes.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;400&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;true&lt;/argument&gt;</span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.RemoveDuplicatedGenes&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_PredictedGenes.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_NoDuplicates.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_DiscardedGenes.xml&lt;/argument&gt;</span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.SolveOverlappings&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_NoDuplicates.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_SolvedOverlaps.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;102&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_RNA_blastn.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_sequences.fna&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;genetic_code.txt&lt;/argument&gt;</span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.GenerateFastaFiles&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_SolvedOverlaps.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_protein_nucleotide_sequences.fasta&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_protein_aminoacid_sequences.fasta&lt;/argument&gt;</span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;  </span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.FillDataFromUniprot&lt;/class_full_name&gt; </span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_SolvedOverlaps.xml&lt;/argument&gt; </span>
<span class="s">      &lt;argument&gt;${name}_Annotation_Dismissed_incl.xml&lt;/argument&gt; </span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">    &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.RemoveDismissedGenes&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Annotation_Dismissed_incl.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Annotation.xml&lt;/argument&gt;</span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.GenerateGffFile&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Annotation.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Annotation.gff&lt;/argument&gt;</span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.GenerateCSVFile&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Annotation.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Annotation.tsv&lt;/argument&gt;</span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.GetIntergenicSequences&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Annotation.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_sequences.fna&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Intergenic.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Intergenic.fasta&lt;/argument&gt;      </span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.gb.ExportGenBankFiles&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_Annotation.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_GenBankExternalData.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_sequences.fna&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}&lt;/argument&gt;     </span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">  &lt;execution&gt;</span>
<span class="s">    &lt;class_full_name&gt;com.era7.bioinfo.annotation.BasicQualityControl&lt;/class_full_name&gt;</span>
<span class="s">    &lt;arguments&gt;</span>
<span class="s">      &lt;argument&gt;${name}_PredictedGenes.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_DiscardedGenes.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_NoDuplicates.xml&lt;/argument&gt;</span>
<span class="s">      &lt;argument&gt;${name}_BasicControlQuality.xml&lt;/argument&gt;     </span>
<span class="s">    &lt;/arguments&gt;</span>
<span class="s">  &lt;/execution&gt;</span>
<span class="s">&lt;/scheduled_executions&gt;</span>
<span class="s">EOF</span>

</pre></div></td></tr><tr><td class=docs>
<p>ugly hack, executions.xml should be a param</p>
</td><td class=code><div class=highlight><pre>
cp <span class="nv">$BG7_HOME</span>/jar/bg7.jar <span class="nv">$output_folder</span>/
</pre></div></td></tr><tr><td class=docs>
<p>run bg7!
I will add memory conf through the script later on, no time for this now</p>
</td><td class=code><div class=highlight><pre>
<span class="nb">echo</span> <span class="s2">&quot;running bg7&quot;</span>
java -d64 -Xmx60G -Xms10G -jar <span class="nv">$output_folder</span>/bg7.jar
</pre></div></td></tr><tr><td class=docs>
<p>clean up</p>
</td><td class=code><div class=highlight><pre>
rm -f <span class="nv">$output_folder</span>/bg7.jar



</pre></div></td></tr><tr><td class=docs>
</td><td class=code><div class=highlight><pre>

</pre></div></td></tr><tr><td class=docs></td><td class='code'></td></tr>
    </tbody>
    </table>
</div>
</body>
</html>
